### PR TITLE
Cart rules country restriction

### PR DESCRIPTION
**Pull request by cfarhani06**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           |  8.0.x 
| Description?      | BO - Cart rules - Case 10 - Country Restriction
| Type?             | refacto
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | no
| Related PRs       | no

### CODE REVIEW

File: tests/UI/pages/BO/quickAccess/index.js [hunk 1]
```diff
@@ -62,6 +63,15 @@ class QuickAccess extends BOBasePage {
     await this.clickAndWaitForNavigation(page, this.addNewQuickAccessButton);
   }
 
+  /**
+   * Go to quick access page
```
Comment (Progi1984) [inline tests/UI/pages/BO/quickAccess/index.js]: The description is not correct.

File: tests/UI/pages/BO/catalog/discounts/add.js [hunk 2]
```diff
@@ -59,6 +59,14 @@ class AddCartRule extends BOBasePage {
     this.quantityPerUserInput = 'input[name=quantity_per_user]';
 
     // Restrictions
+    // Country Group Selection
+    this.countrySelection = '#country_restriction';
+    this.countryGroupSelection = '#country_select_2';
+    this.firstCountrySelection = `${this.countryGroupSelection} option:nth-child(1)`;
+    this.secondCountrySelection = `${this.countryGroupSelection} option:nth-child(2)`;
```
Comment (Progi1984) [inline tests/UI/pages/BO/catalog/discounts/add.js]:
```suggestion
    this.itemCountrySelection = item => `${this.countryGroupSelection} option:nth-child($[item})`;
    this.firstCountrySelection = this.itemCountrySelection(1);
    this.secondCountrySelection = this.itemCountrySelection(2);
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+      await countriesPage.filterTable(page, 'input', 'b!name', countries.unitedStates.name);
+
+      const numberOfCountriesAfterFilter = await countriesPage.getNumberOfElementInGrid(page);
+      await expect(numberOfCountriesAfterFilter).to.be.equal(1);
+
+      const textColumn = await countriesPage.getTextColumnFromTable(page, 1, 'b!name', countries.afghanistan.name);
+      await expect(textColumn).to.equal(countries.unitedStates.name);
+    });
+
+    it('should check that this country is not activated\'', async function () {
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
    it('should check that this country is not activated', async function () {
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+        baseContext,
+      );
+
+      await cartPage.addPromoCode(page, cartRuleCode.code);
+
+      const chooseDeliveryAddressNotification = await cartPage.getAlertWarningForDeliveryAdress(page);
+      await expect(chooseDeliveryAddressNotification).to.equal(cartPage.alertChooseDeliveryAddressWarningtext);
+    });
+
+    it('should proceed to checkouts', async function () {
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
    it('should proceed to checkout', async function () {
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+      // Proceed to checkout the shopping cart
+      await cartPage.clickOnProceedToCheckout(page);
+
+      const isCheckout = await checkoutPage.isCheckoutPage(page);
+      await expect(isCheckout).to.be.true;
+    });
+
+    it('should checkout by signIn', async function () {
+      await testContext.addContextItem(this, 'testIdentifier', 'SignInFO', baseContext);
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
      await testContext.addContextItem(this, 'testIdentifier', 'signInFO', baseContext);
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+
+      const isCustomerConnected = await checkoutPage.customerLogin(page, DefaultCustomer);
+      await expect(isCustomerConnected, 'Customer is not connected').to.be.true;
+    });
+
+    it('should confirm adress after signIn', async function () {
+      await testContext.addContextItem(
+        this,
+        'testIdentifier',
+        'ChooseAndConfirmAdressStep',
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
        'chooseAndConfirmAdressStep',
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+        baseContext,
+      );
+
+      await checkoutPage.chooseDeliveryAddress(page, 2);
+
+      const isDeliveryStep = await checkoutPage.goToDeliveryStep(page);
+      await expect(isDeliveryStep, 'Delivery Step boc is not displayed').to.be.true;
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
      await expect(isDeliveryStep, 'Delivery Step block is not displayed').to.be.true;
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+    });
+
+    it('should set the promo code for second time and check total after discount', async function () {
+      await testContext.addContextItem(
+        this,
+        'testIdentifier',
+        'addPromoCodeANDVerifyTotalAfterDiscount',
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
        'addPromoCodeAndVerifyTotalAfterDiscount',
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+
+      const chooseDeliveryAddressNotification = await cartPage.getAlertWarningForDeliveryAdress(page);
+      await expect(chooseDeliveryAddressNotification).to.equal(cartPage.alertChooseDeliveryAddressWarningtext);
+    });
+
+    it('should proceed to checkouts', async function () {
+      await testContext.addContextItem(
+        this,
+        'testIdentifier',
+        'ProceedToCheckoutAndSignIn',
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
        'proceedToCheckoutAndSignIn',
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+      const notificationsNumber = await cartPage.getCartNotificationsNumber(page);
+      await expect(notificationsNumber).to.be.equal(1);
+    });
+
+    it('should set the promo code and verify the error message', async function () {
+      await testContext.addContextItem(
+        this,
+        'testIdentifier',
+        'AddPromoCode',
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
        'addPromoCode',
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+
+      const pageTitle = await foProductPage.getPageTitle(page);
+      await expect(pageTitle.toUpperCase()).to.contains(ProductData.firstProductData.name);
+    });
+
+    it('should add product to cart and proceed to checkout', async function () {
+      await testContext.addContextItem(
+        this,
+        'testIdentifier',
+        'AddProductToCart',
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
        'addProductToCart',
```

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+
+      const isHomePage = await foHomePage.isHomePage(page);
+      await expect(isHomePage, 'Fail to open FO home page').to.be.true;
+    });
+
+    it('should go to the first product page', async function () {
+      await testContext.addContextItem(
+        this,
+        'testIdentifier',
+        'GoToFirstProductPage',
```
Comment (Progi1984) [inline tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]:
```suggestion
        'goToFirstProductPage',
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks :)

File: tests/UI/pages/BO/quickAccess/index.js (hunk 1 above)
Comment (cfarhani06) [reply tests/UI/pages/BO/quickAccess/index.js]: Done

File: tests/UI/pages/BO/catalog/discounts/add.js (hunk 2 above)
Comment (cfarhani06) [reply tests/UI/pages/BO/catalog/discounts/add.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 3 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 4 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 5 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 6 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 7 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 8 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 9 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 10 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 11 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

File: tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js (hunk 12 above)
Comment (cfarhani06) [reply tests/UI/campaigns/functional/BO/03_catalog/07_discounts/01_cartRules/01_CRUDCartRule/10_countryRestriction.js]: Done

**Comment by nesrineabdmouleh** [comment]
Hello @cfarhani06,
Please check the CI error (UI tests code checks / Checking doubles in steps identifiers)
Thank you!

**Comment by Progi1984** [comment]
It seems that Github detects ea0126be7a28678be248b3a07717f2266e8f7d7b as a merge commit.

**Comment by cfarhani06** [comment]
Another [PR](https://github.com/PrestaShop/PrestaShop/pull/30049) was created for the same scenario

**Pull request closed without merging**